mlflow/xgboost.py

```diff
@@ -286,6 +288,24 @@ def autolog(importance_types=["weight"]):  # pylint: disable=W0102
     import xgboost
     import numpy as np
 
+    def __init__(self, *args, **kwargs):
+        data = args[0] if len(args) > 0 else kwargs["data"]
+
+        original = gorilla.get_original_attribute(xgboost.DMatrix, "__init__")
+
+        data_copy = None
+        try:
+            if isinstance(data, str):
+                raise Exception("The input data was of type string.")
+
+            data_copy = deepcopy(data[:5])
+        except Exception as e:
+            msg = "Failed to gather an input example: " + str(e)
```

Comment:
It seems a bit odd that we're logging a warning message when the dataset is constructed rather than during the training process. Users may construct dataset objects outside of the training context; e.g., they may create an XGBoost dataset and serialize it in binary format to be shared / reloaded later. In that case, logging a warning message such as "Failed to gather an input example" may seem quite confusing.

Instead, perhaps we could hold on to the exception string via the `data_copy` attribute and use it to populate the warning message emitted in `train()` when an input example cannot be found? We could use an internal container object (e.g., `_InputExampleInfo`) which contains `input_example` and `error` fields that are populated in a mutually exclusive fashion.